### PR TITLE
Valentyn/an 1064 base traversal strategy

### PR DESCRIPTION
**Pull request by vkagamlyk**
*Issue #, if available:*
https://bitquill.atlassian.net/browse/AN-1064

*Description of changes:*
Implement base TraversalStrategy

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/driver/strategies_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		assert.NotNil(t, count)
+		assert.Equal(t, 1, len(count))
+		val, err := count[0].GetInt32()
+		assert.Equal(t, int32(6), val)
+	})
+
+	t.Run("Test write with TestReadOnlyStrategy", func(t *testing.T) {
+		g := initializeGraph(t, testNoAuthUrl, &AuthInfo{}, &tls.Config{})
+
+		_, err := g.WithStrategies(ReadOnlyStrategy()).AddV("person").Property("name", "foo").ToList()
```
Comment (lyndonbauto) [inline gremlin-go/driver/strategies_test.go:45]:
```suggestion
		_, promise, err := g.WithStrategies(ReadOnlyStrategy()).AddV("person").Property("name", "foo").Iterate()
		assert.NotNil(t, err)
		<-promise
```

File: gremlin-go/driver/strategies_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+		assert.NotNil(t, err)
```
Comment (lyndonbauto) [inline gremlin-go/driver/strategies_test.go]:
```suggestion
		```suggestion
		_, promise, err := g.WithStrategies(ReadOnlyStrategy()).AddV("person").Property("name", "foo").Iterate()
		assert.NotNil(t, err)
		<-promise
```
```

Comment (vkagamlyk) [reply gremlin-go/driver/strategies_test.go]: fixed in https://github.com/lyndonbauto/tinkerpop/pull/55/commits/fe5bc1697327acf9c284efd75c15af1edd7190ca

**Review by lyndonbauto — APPROVED** [review]

File: gremlin-go/driver/traversal.go [hunk 3]
```diff
@@ -83,13 +83,13 @@ func (t *Traversal) Iterate() (*Traversal, <-chan bool, error) {
...
 
-	r := make(chan bool)
+	r := make(chan error)
 	go func() {
 		defer close(r)
 
 		// Force waiting until complete.
-		_, _ = res.All()
-		r <- true
+		_, err = res.All()
```
Comment (xiazcy) [inline gremlin-go/driver/traversal.go:91]: For testing, or for checking in general, would we need to assert these error items in the channel are all nil?

Comment (lyndonbauto) [reply gremlin-go/driver/traversal.go:91]: Yeah we should actually. Perhaps add assert.NotNil(<-promise) to other test cases that use Iterate().

**Review by L0Lmaker — APPROVED** [review]

File: gremlin-go/driver/traversal.go (hunk 3 above)
Comment (vkagamlyk) [reply gremlin-go/driver/traversal.go:91]: added in https://github.com/lyndonbauto/tinkerpop/pull/55/commits/5a1d2690caf649aff5e13240f86199a708006ffe

**Review by xiazcy — CHANGES_REQUESTED** [review]
Since we now have remote strategies, should we also take `remote *DriverRemoteConnection` out of the `Traversal` struct (which is not in python Traversal class), and wire the remote strategy into the Traversal functions?

File: gremlin-go/driver/traversal.go [hunk 4]
```diff
@@ -67,7 +67,7 @@ func (t *Traversal) ToSet() (map[*Result]bool, error) {
 }
 
 // Iterate all the Traverser instances in the traversal and returns the empty traversal.
-func (t *Traversal) Iterate() (*Traversal, <-chan bool, error) {
+func (t *Traversal) Iterate() (*Traversal, <-chan error, error) {
 	// TODO: AN-979 This wont be needed once DriverRemoteConnection is replaced by TraversalStrategy
```
Comment (xiazcy) [inline gremlin-go/driver/traversal.go]:
```suggestion
	// TODO: AN-1065 This wont be needed once DriverRemoteConnection is replaced by TraversalStrategy
```
Please also change the ticket number in the comment to AN-1065 for line 43.

**Review by xiazcy — APPROVED** [review]
Double checked and saw that we'll be editing the wiring in AN-1065. So will approve this one.

File: gremlin-go/driver/traversal.go (hunk 4 above)
Comment (vkagamlyk) [reply gremlin-go/driver/traversal.go]: done in https://github.com/lyndonbauto/tinkerpop/pull/55/commits/b0c36e9d210b92c91dee4436024fcce0edbeb1ac

**Pull request merged**